### PR TITLE
Add GET and POST /api/v2/smart-contracts/:address_hash/audit-reports

### LINKED ISSUE #9030: API for submitting audit information about contract
Add API endpoint for submitting audit information about smart-contract.
Multiple audit reports should be available to submit through API. A single contract can have multiple audits.

Audit should contain:
- auditor name
- link to the audit report
- audit date
- memo
- link to the contract address

We should also ask info about audit submitter:
- name
- email address
- project name
- project website

API endpoint should be cpatcha-protected.
Publishing of the audit is supposed to be manual (e.g. through `published` flag).

### PR DESCRIPTION
**Pull request by nikitosing**
Close #9030 

## Changelog
- Add GET and POST `/api/v2/smart-contracts/:address_hash/audit-reports`
- New envs: 
  - `SMART_CONTRACT_AUDIT_REPORTS_AIRTABLE_URL`
  - `SMART_CONTRACT_AUDIT_REPORTS_AIRTABLE_API_KEY`
 https://github.com/blockscout/docs/pull/221
## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/explorer/priv/repo/migrations/20231229120232_add_smart_contract_audit_reports_table.exs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      add(:project_url, :string, null: false)
+
+      add(:audit_company_name, :string, null: false)
+      add(:audit_report_url, :string, null: false)
+      add(:audit_publish_date, :date, null: false)
+
+      add(:request_id, :string, null: true)
+
+      add(:comment, :text, null: true)
+    end
```
Comment (vbaranov) [inline apps/explorer/priv/repo/migrations/20231229120232_add_smart_contract_audit_reports_table.exs:23]: I would suggest adding inserted_at and updated_at timestamps.

**Review by vbaranov — COMMENTED** [review]
It seems that:
- there is no validation on `audit_report_url` is a valid URL
- only letters should be allowed in `submitter_name` value
- we shouldn't insert a record of the audit report into the DB, if Airtable record wasn't created. In my case Airtable record wasn't created, API returned:
```
{
    "errors": {
        "smart_contract_address_hash": [
            "AirTable error. Please try again later"
        ]
    }
}
```
but DB record appeared.
- we should handle such DB errors:
```
Server: localhost:4000 (http)
Request: POST /api/v2/smart-contracts/0x9aE6EF3bFf74F684D22d4541314E61e11Ce8bC98/audit-reports
** (exit) an exception was raised:
    ** (Postgrex.Error) ERROR 22001 (string_data_right_truncation) value too long for type character varying(255)
        (ecto_sql 3.11.1) lib/ecto/adapters/sql.ex:1054: Ecto.Adapters.SQL.raise_sql_call_error/1
        (ecto 3.11.1) lib/ecto/repo/schema.ex:775: Ecto.Repo.Schema.apply/4
        (ecto 3.11.1) lib/ecto/repo/schema.ex:377: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
        (explorer 6.0.0) lib/explorer/chain/smart_contract/audit_report.ex:81: Explorer.Chain.SmartContract.AuditReport.create/1
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:268: BlockScoutWeb.API.V2.SmartContractController.audit_report_submission/2
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:1: BlockScoutWeb.API.V2.SmartContractController.action/2
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:1: BlockScoutWeb.API.V2.SmartContractController.phoenix_controller_pipeline/2
        (phoenix 1.5.14) lib/phoenix/router.ex:352: Phoenix.Router.__call__/2
        (phoenix 1.5.14) lib/phoenix/router/route.ex:41: Phoenix.Router.Route.call/2
        (phoenix 1.5.14) lib/phoenix/router.ex:352: Phoenix.Router.__call__/2
        (phoenix 1.5.14) lib/phoenix/router/route.ex:41: Phoenix.Router.Route.call/2
        (phoenix 1.5.14) lib/phoenix/router.ex:352: Phoenix.Router.__call__/2
```
- I was able to create an audit report with `audit_publish_date` in the future.
- I was able to insert audit reports for the same address with the same audit_: report_url/publish_date/company_name. I suppose this set should be unique.

In order to prevent spamming of this table in addition to captcha, could we also add limit for the number of audits per day for the contract address?

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex [hunk 2]
```diff
@@ -247,4 +299,13 @@ defmodule BlockScoutWeb.API.V2.SmartContractController do
...
   def prepare_args(other), do: [other]
+
+  defp validate_smart_contract(params, address_hash_string) do
+    with {:format, {:ok, address_hash}} <- {:format, Chain.string_to_address_hash(address_hash_string)},
+         {:ok, false} <- AccessHelper.restricted_access?(address_hash_string, params),
+         {:not_found, smart_contract} when not is_nil(smart_contract) <-
+           {:not_found, SmartContract.address_hash_to_smart_contract(address_hash, @api_true)} do
+      {:ok, address_hash, smart_contract}
+    end
+  end
```
Comment (sl1depengwyn) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:310]: It seems that we can use this function in `solidityscan_report`, `query_read_method`, `methods_write_proxy`, `methods_read_proxy`, `methods_write`, `methods_read` and maybe in `smart_contract` I'm not sure

I suggest to do it in this pr, it seems reasonable for me as we introduce this fucntion in this pr

File: apps/explorer/lib/explorer/third_party_integrations/airtable.ex [hunk 3]
```diff
@@ -6,9 +6,15 @@ defmodule Explorer.ThirdPartyIntegrations.AirTable do
 
   alias Ecto.Changeset
   alias Explorer.Account.PublicTagsRequest
+  alias Explorer.Chain.SmartContract.AuditReport
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/third_party_integrations/airtable.ex:9]: Let's update doc like this
```
    Module is responsible for submitting requests for public tags and audit reports to AirTable
```

**Review by vbaranov — COMMENTED** [review]
@nikitosing 

could we also handle this error:
```
2024-01-25T14:23:14.437 [error] #PID<0.1805.0> running BlockScoutWeb.Endpoint (connection #PID<0.1704.0>, stream id 10) terminated
Server: localhost:4000 (http)
Request: POST /api/v2/smart-contracts/0x9aE6EF3bFf74F684D22d4541314E61e11Ce8bC98/audit-reports
** (exit) an exception was raised:
    ** (CaseClauseError) no case clause matching: []
        (hackney 1.20.1) .../blockscout/deps/hackney/src/hackney_url.erl:248: :hackney_url.parse_netloc/2
        (hackney 1.20.1) .../blockscout/deps/hackney/src/hackney.erl:351: :hackney.request/5
        (httpoison 2.2.1) lib/httpoison/base.ex:888: HTTPoison.Base.request/6
        (explorer 6.0.0) lib/explorer/third_party_integrations/airtable.ex:79: Explorer.ThirdPartyIntegrations.AirTable.submit_entry/4
        (explorer 6.0.0) lib/explorer/chain/smart_contract/audit_report.ex:124: Explorer.Chain.SmartContract.AuditReport.create/1
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:260: BlockScoutWeb.API.V2.SmartContractController.audit_report_submission/2
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:1: BlockScoutWeb.API.V2.SmartContractController.action/2
        (block_scout_web 6.0.0) lib/block_scout_web/controllers/api/v2/smart_contract_controller.ex:1: BlockScoutWeb.API.V2.SmartContractController.phoenix_controller_pipeline/2
```

when both CONTRACT_AUDIT_REPORTS_AIRTABLE_URL and CONTRACT_AUDIT_REPORTS_AIRTABLE_API_KEY are not provided

**Comment by nikitosing** [comment]
@vbaranov fixed in e4a974439a8d15356d52a35d9ceddb8f88a60b47

**Review by vbaranov — APPROVED** [review]

**Review by sl1depengwyn — APPROVED** [review]

**Pull request merged**